lib/TaskProcessing/TextToTextChatProvider.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OCA\OpenAi\TaskProcessing;
+
+use Exception;
+use OCA\OpenAi\AppInfo\Application;
+use OCA\OpenAi\Service\OpenAiAPIService;
+use OCP\IConfig;
+use OCP\IL10N;
+use OCP\TaskProcessing\EShapeType;
+use OCP\TaskProcessing\ISynchronousProvider;
+use OCP\TaskProcessing\ShapeDescriptor;
+use OCP\TaskProcessing\TaskTypes\TextToTextChat;
+use RuntimeException;
+
+class TextToTextChatProvider implements ISynchronousProvider {
+
+	public function __construct(
+		private OpenAiAPIService $openAiAPIService,
+		private IConfig $config,
+		private IL10N $l,
+		private ?string $userId,
+	) {
+	}
+
+	public function getId(): string {
+		return Application::APP_ID . '-text2text:chat';
+	}
+
+	public function getName(): string {
+		return $this->openAiAPIService->getServiceName();
+	}
+
+	public function getTaskTypeId(): string {
+		return TextToTextChat::ID;
+	}
+
+	public function getExpectedRuntime(): int {
+		return $this->openAiAPIService->getExpTextProcessingTime();
+	}
+
+	public function getOptionalInputShape(): array {
+		return [
+			'max_tokens' => new ShapeDescriptor(
+				$this->l->t('Maximum output words'),
+				$this->l->t('The maximum number of words/tokens that can be generated in the completion.'),
+				EShapeType::Number
+			),
+		];
+	}
+
+	public function getOptionalOutputShape(): array {
+		return [];
+	}
+
+	public function process(?string $userId, array $input, callable $reportProgress): array {
+		$startTime = time();
+		$adminModel = $this->config->getAppValue(Application::APP_ID, 'default_completion_model_id', Application::DEFAULT_COMPLETION_MODEL_ID) ?: Application::DEFAULT_COMPLETION_MODEL_ID;
+
+		if (!isset($input['input']) || !is_string($input['input'])) {
+			throw new RuntimeException('Invalid input');
+		}
+		$userPrompt = $input['input'];
+
+		if (!isset($input['system_prompt']) || !is_string($input['system_prompt'])) {
+			throw new RuntimeException('Invalid system_prompt');
+		}
+		$systemPrompt = $input['system_prompt'];
+
+		if (!isset($input['history']) || !is_array($input['history'])) {
+			throw new RuntimeException('Invalid history');
+		}
+		$history = $input['history'];
+
+		$maxTokens = null;
+		if (isset($input['max_tokens']) && is_int($input['max_tokens'])) {
+			$maxTokens = $input['max_tokens'];
+		}
+
+		try {
+			$completion = $this->openAiAPIService->createChatCompletion($this->userId, $adminModel, $userPrompt, $systemPrompt, $history, 1, $maxTokens);
```

Comment:
Does this take into account the roles? ie. that the role of `history[0]` is `user`?